@mwesterby Looks like we have a checkstyle issue:
```
Task :server:checkstyleTest

[ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-14942/server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java:336:81: '{' is not followed by whitespace. [WhitespaceAround]

[ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-14942/server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java:336:82: '}' is not preceded with whitespace. [WhitespaceAround]

[ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-14942/server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java:344:81: '{' is not followed by whitespace. [WhitespaceAround]

[ant:checkstyle] [ERROR] /home/jenkins/jenkins-agent/workspace/Kafka_kafka-pr_PR-14942/server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java:344:82: '}' is not preceded with whitespace. [WhitespaceAround]



> Task :server:checkstyleTest FAILED
```

--- a/build.gradle
+++ b/build.gradle
@@ -849,6 +849,7 @@ project(':server') {
   dependencies {
     implementation project(':clients')
     implementation project(':server-common')
+    implementation libs.metrics
 
     implementation libs.slf4jApi
 

--- a/checkstyle/import-control-server.xml
+++ b/checkstyle/import-control-server.xml
@@ -58,6 +58,11 @@
   <!-- utilities and reusable classes from server-common -->
   <allow pkg="org.apache.kafka.queue" />
   <allow pkg="org.apache.kafka.server.common" />
+  <allow pkg="org.apache.kafka.server.metrics" />
+  <allow pkg="com.yammer.metrics" />
+
+  <!-- utilities for testing -->
+  <allow pkg="org.apache.kafka.test" />
 
   <!-- persistent collection factories/non-library-specific wrappers -->
   <allow pkg="org.apache.kafka.server.immutable" exact-match="true" />

--- a/server/src/main/java/org/apache/kafka/server/AssignmentsManager.java
+++ b/server/src/main/java/org/apache/kafka/server/AssignmentsManager.java
@@ -17,6 +17,7 @@
 
 package org.apache.kafka.server;
 
+import com.yammer.metrics.core.Gauge;
 import org.apache.kafka.clients.ClientResponse;
 import org.apache.kafka.common.Uuid;
 import org.apache.kafka.common.message.AssignReplicasToDirsRequestData;
@@ -34,6 +35,7 @@ import org.apache.kafka.common.utils.Utils;
 import org.apache.kafka.queue.EventQueue;
 import org.apache.kafka.queue.KafkaEventQueue;
 import org.apache.kafka.server.common.TopicIdPartition;
+import org.apache.kafka.server.metrics.KafkaMetricsGroup;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -67,15 +69,20 @@ public class AssignmentsManager {
 
     private static final long MAX_BACKOFF_INTERVAL_MS = TimeUnit.SECONDS.toNanos(10);
 
+    // visible for testing.
+    static final String QUEUE_REPLICA_TO_DIR_ASSIGNMENTS_METRIC_NAME = "QueuedReplicaToDirAssignments";
+
     private final Time time;
     private final NodeToControllerChannelManager channelManager;
     private final int brokerId;
     private final Supplier<Long> brokerEpochSupplier;
     private final KafkaEventQueue eventQueue;
+    private final KafkaMetricsGroup metricsGroup = new KafkaMetricsGroup(this.getClass());
 
-    // These variables should only be mutated from the KafkaEventQueue thread
-    private Map<TopicIdPartition, AssignmentEvent> inflight = null;
-    private Map<TopicIdPartition, AssignmentEvent> pending = new HashMap<>();
+    // These variables should only be mutated from the KafkaEventQueue thread,
+    // but `inflight` and `pending` are also read from a Yammer metrics gauge.
+    private volatile Map<TopicIdPartition, AssignmentEvent> inflight = null;
+    private volatile Map<TopicIdPartition, AssignmentEvent> pending = new HashMap<>();
     private final ExponentialBackoff resendExponentialBackoff =
             new ExponentialBackoff(100, 2, MAX_BACKOFF_INTERVAL_MS, 0.02);
     private int failedAttempts = 0;
@@ -93,10 +100,24 @@ public class AssignmentsManager {
                 "broker-" + brokerId + "-directory-assignments-manager-",
                 new ShutdownEvent());
         channelManager.start();
+        this.metricsGroup.newGauge(QUEUE_REPLICA_TO_DIR_ASSIGNMENTS_METRIC_NAME, new Gauge<Integer>() {
+            @Override
+            public Integer value() {
+                return getMapSize(inflight) + getMapSize(pending);
+            }
+
+            private int getMapSize(Map<TopicIdPartition, AssignmentEvent> map) {
+                return map == null ? 0 : map.size();
+            }
+        });
     }
 
     public void close() throws InterruptedException {
-        eventQueue.close();
+        try {
+            eventQueue.close();
+        } finally {
+            metricsGroup.removeMetric(QUEUE_REPLICA_TO_DIR_ASSIGNMENTS_METRIC_NAME);
+        }
     }
 
     public void onAssignment(TopicIdPartition topicPartition, Uuid dirId, Runnable callback) {

--- a/server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java
+++ b/server/src/test/java/org/apache/kafka/server/AssignmentsManagerTest.java
@@ -17,6 +17,9 @@
 
 package org.apache.kafka.server;
 
+import com.yammer.metrics.core.Gauge;
+import com.yammer.metrics.core.Metric;
+import com.yammer.metrics.core.MetricName;
 import org.apache.kafka.clients.ClientResponse;
 import org.apache.kafka.common.Uuid;
 import org.apache.kafka.common.errors.AuthenticationException;
@@ -29,6 +32,8 @@ import org.apache.kafka.common.requests.AssignReplicasToDirsResponse;
 import org.apache.kafka.common.utils.MockTime;
 import org.apache.kafka.metadata.AssignmentsHelper;
 import org.apache.kafka.server.common.TopicIdPartition;
+import org.apache.kafka.server.metrics.KafkaYammerMetrics;
+import org.apache.kafka.test.TestUtils;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -303,5 +308,41 @@ public class AssignmentsManagerTest {
             time.sleep(TimeUnit.SECONDS.toMillis(1));
             manager.wakeup();
         }
+    }
+
+    static Metric findMetric(String name) {
+        for (Map.Entry<MetricName, Metric> entry : KafkaYammerMetrics.defaultRegistry().allMetrics().entrySet()) {
+            MetricName metricName = entry.getKey();
+            if (AssignmentsManager.class.getSimpleName().equals(metricName.getType()) && metricName.getName().equals(name)) {
+                return entry.getValue();
+            }
+        }
+        throw new IllegalArgumentException("metric named " + name + " not found");
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    void testQueuedReplicaToDirAssignmentsMetric() throws Exception {
+        CountDownLatch readyToAssert = new CountDownLatch(1);
+        doAnswer(invocation -> {
+            readyToAssert.countDown();
+            return null;
+        }).when(channelManager).sendRequest(any(AssignReplicasToDirsRequest.Builder.class), any(ControllerRequestCompletionHandler.class));
+
+        Gauge<Integer> queuedReplicaToDirAssignments = (Gauge<Integer>) findMetric(AssignmentsManager.QUEUE_REPLICA_TO_DIR_ASSIGNMENTS_METRIC_NAME);
+        assertEquals(0, queuedReplicaToDirAssignments.value());
+
+        for (int i = 0; i < 4; i++) {
+            manager.onAssignment(new TopicIdPartition(TOPIC_1, i), DIR_1, () -> { });
+        }
+        while (!readyToAssert.await(1, TimeUnit.MILLISECONDS)) {
+            time.sleep(100);
+        }
+        assertEquals(4, queuedReplicaToDirAssignments.value());
+
+        for (int i = 4; i < 8; i++) {
+            manager.onAssignment(new TopicIdPartition(TOPIC_1, i), DIR_1, () -> { });
+        }
+        TestUtils.retryOnExceptionWithTimeout(5_000, () -> assertEquals(8, queuedReplicaToDirAssignments.value()));
     }
 }

